src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+		case 0x84 ... 0xef: /* HID KBD code, release */
+			type = RATBAG_MACRO_EVENT_KEY_RELEASED;
+			event_data = ratbag_hidraw_get_keycode_from_keyboard_usage(
+			    device, data[i] - 0x80);
+			break;
+		case 0xf0 ... 0xfc: /* KBD modifier/mouse button release */
+			type = RATBAG_MACRO_EVENT_KEY_RELEASED;
+			event_data = gskill_misc_macro_code_to_keycode(data[i]);
+			break;
+		case 0xfe:
+			type = RATBAG_MACRO_EVENT_KEY_PRESSED;
+			event_data = KEY_SCROLLUP;
+			break;
+		}
+
+		ratbag_button_macro_set_event(macro, event_idx, type,
+					      event_data);
+	}
+
+	return macro;
+}
+
+/* FIXME: the macro struct here should be a const, but it looks like there's a
+ * couple of functions in ratbag that need to have a const qualifier added to
+ * their function declarations
+ */
+static struct gskill_macro_report *
+gskill_macro_to_report(struct ratbag_device *device,
+		       struct ratbag_button_macro *macro,
+		       unsigned int profile, unsigned int button)
+{
+	struct gskill_data *drv_data = ratbag_get_drv_data(device);
+	struct gskill_macro_report *report =
+		&drv_data->profile_data[profile].macros[button];
+	struct gskill_macro_delay *delay;
+	unsigned int event_num = ratbag_button_macro_get_num_events(macro);
+	struct ratbag_macro_event *event;
+	uint8_t *buf = report->macro_content;
+	int profile_pos, increment, event_idx;
+
+	memset(report, 0, sizeof(*report));
+	strcpy(macro->macro.name, report->macro_name);
+	report->macro_name_length = strlen(macro->macro.name);
+
+	report->macro_num = (profile * 10) + button;
+	report->macro_exec_method = GSKILL_MACRO_METHOD_BUTTON_PRESS;
+	report->loop_count = 0;
+	report->please_set_me_to_1_thank_you = 1; /* No prob! Happy to help :) */
+
+	for (profile_pos = 0, increment = 1, event_idx = 0;
+	     event_idx < (signed)event_num;
+	     event_idx++, profile_pos += increment, increment = 1) {
+		event = &macro->macro.events[event_idx];
+
+		switch (event->type) {
+		case RATBAG_MACRO_EVENT_WAIT:
+			delay = (struct gskill_macro_delay*)&buf[profile_pos];
+			increment = sizeof(*delay);
+
+			delay->tag = 1;
+			delay->count = event->event.timeout;
+			break;
+		case RATBAG_MACRO_EVENT_KEY_PRESSED:
+		case RATBAG_MACRO_EVENT_KEY_RELEASED:
+			buf[profile_pos] = gskill_macro_code_from_event(device,
+									event);
+			break;
+		case RATBAG_MACRO_EVENT_INVALID:
+		case RATBAG_MACRO_EVENT_NONE:
+			goto fin;
+		}
+	}
+
+fin:
+	report->macro_length = profile_pos;
+
+	return report;
+}
+
+static struct gskill_macro_report *
+gskill_read_button_macro(struct ratbag_device *device,
+			 unsigned int profile, unsigned int button)
+{
+	struct gskill_data *drv_data = ratbag_get_drv_data(device);
+	struct gskill_macro_report *report =
+		&drv_data->profile_data[profile].macros[button];
+	uint8_t checksum;
+	uint8_t macro_num = (profile * 10) + button;
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x0b, macro_num, 0x0 };
+	int rc;
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD, buf,
```

Comment:
You'd better use gskill_general_cmd... (or maybe a wrapper if you are not interested in checking the error returned by the mouse).